File: src/main/java/chapter5/schedulers/TrampolineSchedulerExample.java
package chapter5.schedulers;

import common.CommonUtils;
import common.Log;
import io.reactivex.Observable;
import io.reactivex.schedulers.Schedulers;


public class TrampolineSchedulerExample {

    public void basic() {
        String[] objs = {"1", "3", "5"};
        Observable<String> source = Observable.fromArray(objs);

        source.subscribeOn(Schedulers.trampoline())
                .map(data -> "<<" + data + ">>")
                .subscribe(Log::it);

        source.subscribeOn(Schedulers.trampoline())
                .map(data -> "##" + data + "##")
                .subscribe(Log::it);
        CommonUtils.sleep(500);
    }

    public static void main(String[] args) {
        TrampolineSchedulerExample demo = new TrampolineSchedulerExample();
        demo.basic();
    }

}

File: src/main/kotlin/chapter5/schedulers/TrampolineSchedulerExample_.kt
package chapter5.schedulers

import common.CommonUtils_
import common.Log_
import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers

class TrampolineSchedulerExample_ {

    fun basic() {
        val objs = arrayOf("1", "3", "5")
        val source = Observable.fromArray(*objs)

        source.subscribeOn(Schedulers.trampoline())
                .map { "<<$it>>" }
                .subscribe { Log_.it(it) }

        source.subscribeOn(Schedulers.trampoline())
                .map { "##$it##" }
                .subscribe { Log_.it(it) }
        CommonUtils_.sleep(500)
    }

}


fun main(args: Array<String>) {
    val demo = TrampolineSchedulerExample_()
    demo.basic()
}
